// smc-project/jupyter/nbconvert.js

/*
Node.js interface to nbconvert.
 */

(function() {
  var defaults, misc, misc_node, required;

  misc = require('smc-util/misc');

  defaults = misc.defaults, required = misc.required;

  misc_node = require('smc-util-node/misc_node');

  exports.nbconvert = function(opts) {
    var args, command, i, j, k, ref, to;
    opts = defaults(opts, {
      args: required,
      directory: void 0,
      timeout: 30,
      cb: required
    });
    to = void 0;
    j = void 0;
    for (i = k = 0, ref = opts.args.length; 0 <= ref ? k < ref : k > ref; i = 0 <= ref ? ++k : --k) {
      if (opts.args[i] === '--to') {
        j = i;
        to = opts.args[i + 1];
        break;
      }
    }
    if (to === 'sagews') {
      command = 'smc-ipynb2sagews';
      args = opts.args.slice(0, j).concat(opts.args.slice(j + 2));
    } else {
      command = 'jupyter';
      args = ['nbconvert'].concat(opts.args);
    }
    return misc_node.execute_code({
      command: command,
      args: args,
      path: opts.directory,
      err_on_exit: true,
      timeout: opts.timeout,
      cb: (function(_this) {
        return function(err, output) {
          if (err && (output != null ? output.stderr : void 0)) {
            err = output != null ? output.stderr : void 0;
          }
          return opts.cb(err);
        };
      })(this)
    });
  };

}).call(this);
